Show latest, minimum and maximum metric values on metric page

adamreeve opened the pull request:
## What changes are proposed in this pull request?

Displays a table of the latest, minimum and maximum metric values in the metric view, as discussed in #5517. This partly addresses #4750. This handles the case when the metric view is used to show data for a single run and when comparing multiple runs.

### Single run

![single-run-multi-metric](https://user-images.githubusercontent.com/626438/161184907-5fa1bea7-8820-471f-9b0e-247726b2cb20.png)

### Multiple runs

![multi-run-single-metric](https://user-images.githubusercontent.com/626438/161184921-f4871934-c929-49fb-9ed1-c70b4dc40c66.png)

### Multiple runs and multiple metrics

![multi-run-multi-metric](https://user-images.githubusercontent.com/626438/161184941-001ac13a-247c-4a5b-90c5-7a8774837b8d.png)

## How is this patch tested?

New unit tests added.

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

## Release Notes

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

The metric view now includes a table showing the latest, minimum and maximum metric values for runs.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [x] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->
<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

adamreeve committed:
Show latest, minimum and maximum metric values on metric page

Signed-off-by: Adam Reeve <adreeve@gmail.com>

adamreeve committed:
Consistent behaviour when no metrics selected

Signed-off-by: Adam Reeve <adreeve@gmail.com>

dbczumar reviewed (commented):
@adamreeve This is really awesome! Tried it out locally and it works great! I have a few small suggestions; let me know what you think of them:

1. Can we turn the `Run <id>` entries in the table into links that users can click on?
2. Can we display the step associated with each metric min / max / latest value in parentheses?

E.g. 

```
Latest            |            Max      |               ....

0.355 (step=17).  |     0.4 (step=15)   |               ....
```

3. Can we make the table columns sortable so that users can easily identify the run with the smallest or largest latest / max / min value for a particular metric?

4. Can we make the tables fixed height and scrollable to conserve page space when there are many metrics being plotted?

Looking forward to shipping this! Thanks so much for your contribution!

@sunishsheth2009 Can you take a look as well and provide feedback?

adamreeve committed:
Show corresponding step numbers next to metric values

Signed-off-by: Adam Reeve <adreeve@gmail.com>

adamreeve committed:
Link to runs when showing metrics for multiple runs

Signed-off-by: Adam Reeve <adreeve@gmail.com>

adamreeve committed:
Make metric summary columns sortable

Signed-off-by: Adam Reeve <adreeve@gmail.com>

adamreeve committed:
Fix metric table height and make it scrollable

Signed-off-by: Adam Reeve <adreeve@gmail.com>

adamreeve committed:
Merge branch 'master' into metric_view_min_max

Signed-off-by: Adam Reeve <adreeve@gmail.com>

adamreeve committed:
Fix after merge: Handle multiple experiment ids in metrics plot panel

Signed-off-by: Adam Reeve <adreeve@gmail.com>

adamreeve committed:
Test for presence of run links

Signed-off-by: Adam Reeve <adreeve@gmail.com>

adamreeve commented:
Thanks for the feedback @dbczumar, all good suggestions and I've had a go at addressing those points. Making the tables fixed height was a bit awkward and meant I had to make the column widths fixed sizes, but I think that's okay and works well now. I also allowed sorting by the run or metric name for completeness, but I'm not sure if that would be used much.

Related to not taking up too much space, I'm also interested in being able to show multiple separate plots on the metric page (https://github.com/mlflow/mlflow/issues/5397#issuecomment-1088281898), so if that's something that might be accepted I'd have to think about how it fits in with this change, eg. making the summary table views collapsible, or maybe having the tables separate to the plots and being able to control the set of metrics in tables separate to the plots. Probably not worth worrying about that too much right now though, I can address that if and when it's needed.

dbczumar reviewed (commented):
@adamreeve Apologies for the delay. This is fantastic! Just a couple of tiny alignment / titling things, then this is ready to go from my perspective. Thank you for your contribution!

1. When there's only one metric selected, can we still show the title for it on the associated table?
2. I noticed that the width of the runs column causes certain run IDs to overflow on to the next line; can we expand the column width slightly?

<img width="315" alt="Screen Shot 2022-04-12 at 2 56 10 PM" src="https://user-images.githubusercontent.com/39497902/163061363-567c9a5e-18e6-4030-a2d6-9b7973174672.png">

@sunishsheth2009 Can you take a look from a code correctness / frontend best practices perspective?

sunishsheth2009 commented on mlflow/server/js/src/experiment-tracking/components/MetricsSummaryTable.js:
```diff
@@ -0,0 +1,242 @@
+    });
+  }
+
+  dataColumns() {
```
Is there a reason that this needs to be a function rather than just a const?

sunishsheth2009 reviewed (commented):
LGTM.

adamreeve committed:
Always show metric name header

Signed-off-by: Adam Reeve <adreeve@gmail.com>

adamreeve committed:
Bump header column width to avoid run ids overflowing

Signed-off-by: Adam Reeve <adreeve@gmail.com>

adamreeve committed:
Test MetricsSummaryTable is rendered by MetricsPlotPanel

Signed-off-by: Adam Reeve <adreeve@gmail.com>

adamreeve commented on mlflow/server/js/src/experiment-tracking/components/MetricsPlotPanel.test.js:
```diff
@@ -414,9 +425,8 @@ describe('unit tests', () => {
   });
   test('should render the number of completed runs correctly', () => {
```
Yep I've added a test for this now

dbczumar reviewed (approved):
LGTM! Thanks so much, @adamreeve !

@sunishsheth2009 Can you give this one more look so that we can get it merged?

sunishsheth2009 reviewed (approved):
LGTM
